shaman/controllers/search.py

```diff
@@ -29,10 +30,52 @@ def index(self, **kw):
         query = self.apply_filters(kw)
         if not query:
             return []
-        return query.order_by(desc(Repo.modified)).all()
+        # order all the results by their modified time, descending (newest first)
+        latest_modified_repos = query.order_by(desc(Repo.modified))
+        distro_list = util.parse_distro_query(kw.get("distros"))
+        if kw.get('sha1', '') == 'latest':
+            seen_sha1s = []
+
+            # go through all the sha1s in the repositories left from the
+            # filtering, skipping the ones already queried for. We don't use
+            # `set` here because it alters the ordering on `modified` from the
+            # initial query
+            for r in latest_modified_repos:
+                if r.sha1 in seen_sha1s:
+                    continue
+                seen_sha1s.append(r.sha1)
+                latest = []
+                if not distro_list:
+                    return latest_modified_repos.filter_by(sha1=r.sha1).all()
+
+                for distro in distro_list:
+                    version_filter = distro["distro_codename"] or distro['distro_version']
+                    latest_repo = latest_modified_repos.filter_by(
+                        sha1=r.sha1,
+                        distro_version=version_filter
+                    ).order_by(desc(Repo.modified)).first()
+                    if latest_repo:
+                        latest.append(latest_repo)
+                if latest:
```

Comment:
You should only return ``latest`` here if ``len(latest) == len(distro_list)`` otherwise you could potentially return a repo for a sha1 that only exists in some of the ``distros`` specified, not all.